forbedret prosent beskrivelse og lager tester

orjanls opened the pull request:
Har lagt inn beskrivelse av funksjonen `prosent`, oppdatert den anonyme funksjonen til å bruke standard R syntaks og lagt til tester.

orjanls committed:
Rettet skrift i dokumentasjon for funksjonen 'num'

orjanls committed:
Rettet 'NULL' til 'numeric()' i test for funksjonen 'num'

orjanls committed:
Rettet og fylt ut dokumentasjon for funksjonen `prosent`

orjanls committed:
Oppdatert anonym funksjon til standard R (4.1.0) syntaks

orjanls committed:
Lagt til tester for funksjonen `prosent`

orjanls committed:
Fjernet unødvendig whitespace

haugedal commented on R/rapporttekst-funksjoner.R:
```diff
@@ -170,9 +170,40 @@ akse_tall_format = function(antall_desimaler = 2, decimal.mark = ",", big.mark =
+#' istedenfor bindestrek («-42»);
+#' og `NA`-verdier blir vist som en kort tankestrek («–»).
+#' 
+#' Tallene blir avrundet til `desimalar` desimalar
```
Når teksten er på bokmål bør «desimalar» nr. 2 vera «desimaler».

statkarl replied:
Blir det ikkje veldig forvirrande at namnet på argumentet er forskjellig på to nærskylde funksjonar (`num()` og `prosent()`), @haugedal?

statkarl replied:
Eg las feil. Det var berre teksten, ikkje argumentnamnet, du ville ha på bokmål, @haugedal. Då er eg einig.

haugedal commented on R/rapporttekst-funksjoner.R:
```diff
@@ -170,9 +170,40 @@ akse_tall_format = function(antall_desimaler = 2, decimal.mark = ",", big.mark =
+#' 
+#' Tallene blir avrundet til `desimalar` desimalar
+#' med den vanlige avrundingsregelen i R.
+#' hVis `desimalar` er NULL,
```
Stor bokstav midt i ordet `hVis`.

haugedal commented on tests/testthat/test-rapporttekst-funksjoner.R:
```diff
@@ -5,7 +5,7 @@ test_that("num() gir ut tankestrek når «x»-argumentet inneholder NA", {
 
 test_that("num() gir ut riktig verdi når kun «x»-argumentet er gitt", {
-  expect_identical(num(NULL), "{\\numprint{}}")
+  expect_identical(num(numeric()), "{\\numprint{}}")
```
Korleis skal `num()` og  `prosent()` fungera med tomme inndata, @statkarl?

statkarl replied:
Som standard for ~alle vektorbaserte funksjonar: Utvektoren skal alltid ha same lengd som innvektoren, så tom innvektor skal gje tom utvektor (av typen tekst).

haugedal replied:
Så då må denne testen (og så funksjonen) endrast?

statkarl replied:
Ja, dei må endrast.

haugedal commented on tests/testthat/test-rapporttekst-funksjoner.R:
```diff
@@ -113,3 +113,33 @@ test_that("akse_tall_format() gir ut en tom tekstvektor hvis den tar
+})
+
+test_that("prosent() gir ut riktig verdi når kun «x»-argumentet er gitt", {
+  expect_identical(prosent(numeric()), logical(0))
```
Korleis skal `num()` og  `prosent()` fungera med tomme inndata, @statkarl?

haugedal replied:
Og denne skal då òg vera `character()`, ikkje `logical()`?

statkarl replied:
Sjå annan kommentar. Det skal vera `character()` her, ikkje `logical()`.

haugedal commented:
Funksjonane bør kanskje sjekka at inndata er tal, og ikkje andre ting. T.d. burde vel ikkje `num("Tekst.")` fungera, men heller gje ei feilmelding. Kan i så fall starta med å legga inn testar der inndata er andre ting enn tal.

Og kva med tal formatert som tekst (`num("12345")`), skal det gje same feilmelding?

statkarl commented:
> Og kva med tal formatert som tekst `num("12345")`, skal det gje same feilmelding?

Viss me har ei feilmelding på alle ikkje-tal, bør det gje feilmelding, ja.

orjanls committed:
Rettet skrivefeil i dokumentasjon for `prosent`

statkarl commented on R/rapporttekst-funksjoner.R:
```diff
@@ -170,9 +170,40 @@ akse_tall_format = function(antall_desimaler = 2, decimal.mark = ",", big.mark =
+#' @param desimalar Antall desimaler som skal vises etter kommaet.
+#' 
+#' @details
+#' Tallene blir konvertert til prosent og blir vist med prosenttegn
+#' («0.1234» blir «12.34 %»);
+#' desimaltegnet blir (dersom språket er norsk)
+#' komma («3,14»),
+#' istedenfor punktum («3.14»);
+#' negative tall blir vist med ekte minustegn («−42»),
+#' istedenfor bindestrek («-42»);
+#' og `NA`-verdier blir vist som en kort tankestrek («–»).
```
Dette stemmer ikkje. Poenget er at ein får ut *LaTeX-kode* for formatering av talet som prosent dersom ein brukar `kvalreg`-klassen vår. Det *kan* stå forklart kva resultatet av LaTeX-koden, men det blir berre ekstrainformasjon.

statkarl replied:
Forresten, det treng ikkje stå forklart her i det heile tatt; ein kan berre visa til hjelpefila for `num()`-funksjonen for detaljar.

statkarl commented on R/rapporttekst-funksjoner.R:
```diff
@@ -215,7 +215,7 @@ akse_tall_format = function(antall_desimaler = 2, decimal.mark = ",", big.mark =
 #' paste0("Andel menn med 2 desimaler er ", prosent(andel_menn, desimalar = 2), ".")
 prosent = function(x, desimalar = 0) {
   prosent_tekst = x |>
-    map_chr(~ num(100 * .x, desimalar) |>
+    map_chr(\(x) num(100 * x, desimalar) |>
```
`map_chr()` er unødvendig her.

haugedal commented:
Funksjonane `num()` og `prosent()` kunne gjerne fått lagt til avsnittet `@return` i dokumentasjonen når me fyrst er i gang.

statkarl commented on R/rapporttekst-funksjoner.R:
```diff
@@ -10,13 +10,13 @@
 #'
-#' Tar inn ein vektor med tal og gjev ut ein ein vektor
+#' Tar inn ein vektor med tal og gjev ut ein vektor
 #' med' LaTeX-kommandoar for finformatert vising av tala,
```
Her er det ein apostrof for mykje.

statkarl commented on R/rapporttekst-funksjoner.R:
```diff
@@ -215,7 +215,7 @@ akse_tall_format = function(antall_desimaler = 2, decimal.mark = ",", big.mark =
-    map_chr(~ num(100 * .x, desimalar) |>
+    map_chr(\(x) num(100 * x, desimalar) |>
       str_c("\\prosent"))
   ifelse(is.na(x), "\\textendash{}", prosent_tekst)
```
Me brukar konsekvent `if_else()`, ikkje `ifelse()`. Førstnemnde er tryggare. (Kunne me fått ein lint-regel på slikt, @haugedal?)

haugedal replied:
Ja, det burde vel gå an å få til, @statkarl.

statkarl commented:
Bør òg fiksa dei mange lintr-klagene, @orjanls.

orjanls committed:
Fjernet unødvendig whitespace

orjanls committed:
Oppdatert `ifelse` til `if_else`

statkarl commented:
@haugedal 
> Funksjonane bør kanskje sjekka at inndata er tal, og ikkje andre ting. T.d. burde vel ikkje `num("Tekst.")` fungera, men heller gje ei feilmelding. Kan i så fall starta med å legga inn testar der inndata er andre ting enn tal.

Det vert ein ny pakkeavhengnad, men kva med å ta i bruk `checkmate`-pakken til slikt. Han verkar veldig kraftig og enkel å bruka, og han støttar mange av dei typiske testane me har for bruk for. Kan for eksempel bruka:

```r
checkmate::assert_numeric(x, finite = TRUE)
```

Sjå vignetten for eit eksempel på kor mykje han kan forenkla:
https://cran.r-project.org/web/packages/checkmate/vignettes/checkmate.html
Synspunkt, @haugedal, @orjanls, @EndreEvjen1?

EndreEvjen1 commented:
> Det vert ein ny pakkeavhengnad, men kva med å ta i bruk `checkmate`-pakken til slikt. Han verkar veldig kraftig og enkel å bruka, og han støttar mange av dei typiske testane me har for bruk for. Kan for eksempel bruka:
> 
> ```r
> checkmate::assert_numeric(x, finite = TRUE)
> ```
> 
> Sjå vignetten for eit eksempel på kor mykje han kan forenkla: https://cran.r-project.org/web/packages/checkmate/vignettes/checkmate.html Synspunkt, @haugedal, @orjanls, @EndreEvjen1?

Synes det ser ut som en god pakke som kan gjøre det lettere for oss, så jeg stemmer for.

orjanls commented:
> Sjå vignetten for eit eksempel på kor mykje han kan forenkla:
> https://cran.r-project.org/web/packages/checkmate/vignettes/checkmate.html
> Synspunkt, @haugedal, @orjanls, @EndreEvjen1?

Såg ut som en nyttig pakke som forenkler, så stemmer for `checkmate`

orjanls committed:
Fjernet en feilskreven apostrof

haugedal commented:
Einig i at `checkmate` ser ut som ein nyttig pakke, som både kan forenkla og gjera kode meir oversiktleg og lettlest. Stemmer for å ta han i bruk eg òg.

orjanls committed:
Endret `logical()` til `character()` i test for `prosent`

orjanls committed:
Test at `num` gir feilmelding om inndata er i tekstform

orjanls committed:
Lagt til `assert_numeric()` for å bare ta inn vektor men tall

orjanls committed:
Rettet test for `num` til å forvente korrekt utdata

orjanls committed:
Tester for at `num` gir ut riktig når `desimalar = NULL`

orjanls committed:
Tester for at `num` ikke konverterer til eksponentiell notasjon

orjanls committed:
`num`-dokumentasjon for `@return` og dersom inndata er `NULL`

orjanls committed:
Forandret `num` til å gi ut en tom tekstvektor dersom innvektor er tom

orjanls committed:
Rettet beskrivelse av `num` og lagt til `@return`

orjanls committed:
Tester for at `prosent` gir error om inndata er i tekstform

orjanls committed:
Fjernet `map_chr` *og* lagt til `assert_numeric` for `prosent`

orjanls committed:
Lagt til `checkmate::funksjon` der pakken har blitt brukt

orjanls committed:
Lagt til importering av `checkmate (>= 1.7.0)`

orjanls committed:
Lagt til `rlang::funksjon` der pakken blir brukt

haugedal commented on R/rapporttekst-funksjoner.R:
```diff
@@ -170,9 +176,25 @@ akse_tall_format = function(antall_desimaler = 2, decimal.mark = ",", big.mark =
+#' Funksjonaliteten til `num` er gitt i [beskrivelsen av funksjonen][num()].
+#'
+#' @return
+#' Vektor med LaTeX-kode
```
Manglar punktum.

haugedal commented on R/rapporttekst-funksjoner.R:
```diff
@@ -10,84 +10,90 @@
-#' 
+#' ikkje bindestrek («-42»),
+#' `NA`-verdiar vert viste som ein kort tankestrek («–»)
+#' og `NULL`-verdiar returnerar ein tom tekstvektor.
```
Det skal vera «returnerer», ikkje «returnerar».

haugedal commented:
> Bør òg fiksa dei mange lintr-klagene, @orjanls.

Whitespace-klagane kan enkelt fiksast ved å køyra styler på filene.

Argumenta `decimal.mark` og `big.mark` er brukt utan snake_case for å vera like argumenta i funksjonen dei vert gjevne vidare til (`scales::number_format()`). Her kan me leggja til `nolint`-kommentarar til linjene:
```
# nolint: object_name_linter
```

Del eventuelt over fleire linjer:
```
akse_tall_format = function(antall_desimaler = 2,
                            decimal.mark = ",", # nolint: object_name_linter
                            big.mark = " ", # nolint: object_name_linter
                            ...) {
```

statkarl commented on tests/testthat/test-rapporttekst-funksjoner.R:
```diff
@@ -113,3 +143,44 @@ test_that("akse_tall_format() gir ut en tom tekstvektor hvis den tar
+  expect_identical(prosent(numeric()), character(0))
+  expect_identical(prosent(0.1234), "{\\numprint{12}}\\prosent")
+  expect_identical(
+    prosent(c(0.1234, 0.31415)),
+    c("{\\numprint{12}}\\prosent", "{\\numprint{31}}\\prosent")
```
Her (og i andre testar) har du hardkoda at `prosent()` skal gje ut `\numprint{}`-kommandoar med formaterte talverdiar. Det står det ingenting i dokumentasjonen om, og det er langt frå sikkert at han skal gjera dette i framtida. Det er altså ein unødvendig og uheldig avhengnad til `num()` sin gjeldande funksjonalitet i testane.

(Akkurat som funksjonar skal vera mest mogleg uavhengig av *gjeldande* åtferd til andre funksjonar, gjeld det same testar. Generelt er det slik at ein funksjon berre er ansvarleg for *si eiga åtferd*, og skal berre vera avhengig av at andre funksjonar gjer jobben sin (kva denne er, er definert i dokumentasjonen), ikkje *korleis* dei gjer jobben sin. Eks. er `num()` sin jobb å gje ut ein kvalreg-klasse-kompatibel LaTeX-kommando for formatering av tal med gitt # desimalar. Om dette skjer via `\numprint{}` eller noko anna, er berre noko `num()` skal veta inni seg, og ikkje dela med andre.)

Det `prosent()` *skal* gjera, er å gje ut ein vektor der elementa er lik resultatet av `num()`, men med `\prosent` hekta på (og at val av desimalar vert vidaresendt til tilsvarande argument i `num()`, og tilsv. for ev. andre argument). Det er *det* testane skal testa. Dette gjer òg at du får enklare (og gjerne færre) testar.

statkarl replied:
Ja, forresten, ein må sjølvsagt òg testa at `NA`-verdiar vert handterte rett. Altså, viss ein har ein `NA`-verdi, som `num()` no gjev ut som `–`, skal `prosent()` òg gje han ut *likt som `num()` gjer* (og altså ikkje som `–\prosent`, slik ei naiv implementering ville gitt).

statkarl replied:
Dette må òg fungera rett viss berre *nokre* av elementa i vektoren er `NA`.

statkarl reviewed (changes requested):
Testane må endrast til å ikkje vera avhengig av korleis `num()` fungerer (men berre av *at* `num()` fungerer), som forklart i kommentaren.

statkarl commented on tests/testthat/test-rapporttekst-funksjoner.R:
```diff
@@ -113,3 +143,44 @@ test_that("akse_tall_format() gir ut en tom tekstvektor hvis den tar
+})
+
+test_that("prosent() gir ut riktig verdi med «desimalar»-argumentet", {
+  expect_identical(
+    prosent(0.1234, desimalar = 0),
+    "{\\numprint{12}}\\prosent"
+  )
+  expect_identical(
+    prosent(0.1234, desimalar = 2),
+    "{\\numprint{12.34}}\\prosent"
+  )
+  expect_identical(
+    prosent(c(0.1234, 0.31), desimalar = 3),
+    c("{\\numprint{12.340}}\\prosent", "{\\numprint{31.000}}\\prosent")
+  )
```
Alle desse linjene/testane kan erstattast med éin test:

```r
expect_identical(
  prosent(0.1234, desimalar = 1),
  paste0(num(12.34, desimalar = 1), "\\prosent")
)
```

Det einaste ein treng testa, er altså at `desimalar`-argumentet vert rett vidaresendt til `num()`-funksjonen og at inndata vert rett muliplisert med 100.

statkarl commented on tests/testthat/test-rapporttekst-funksjoner.R:
```diff
@@ -113,3 +143,44 @@ test_that("akse_tall_format() gir ut en tom tekstvektor hvis den tar
+test_that("prosent() gir error om «x»-argument er i tekstform", {
+  expect_error(
+    prosent("0.12"),
+    "Assertion on 'x' failed: Must be of type 'numeric', not 'character'."
```
Her *antar* du noko om korleis checkmate-funksjonen skal fungera og nøyaktig formulering av feilmeldinga. Det har du ingenting med! Det er det berre checkmate-funksjonen sjølv som skal halda orden på.

Kvar funksjonen har ansvar for si eiga åtferd. Bruk heller:

```r
checkmate::check_numeric("0.12")
```

for å få ut feilmeldinga som checkmate-funksjonen skal gje ut når inndata er tekst men skal vera numerisk (denne feilmeldinga kan godt endrast i framtidige versjonar av checkmate).

statkarl reviewed (changes requested):
Nok eit «pass dine eigne saker»-problem.

statkarl commented on tests/testthat/test-rapporttekst-funksjoner.R:
```diff
@@ -26,6 +26,36 @@ test_that("num() gir ut riktig verdi med «desimalar»-argumentet", {
+})
+
+test_that("num() gir error om «x»-argument er i tekstform", {
+  expect_error(
+    num("12"),
+    "Assertion on 'x' failed: Must be of type 'numeric', not 'character'."
+  )
+  expect_error(
+    num(c(12, "3.14")),
+    "Assertion on 'x' failed: Must be of type 'numeric', not 'character'."
+  )
```
Treng ikkje *to* testar som testar akkurat det same. Det held med éin.

statkarl commented on tests/testthat/test-rapporttekst-funksjoner.R:
```diff
@@ -26,6 +26,36 @@ test_that("num() gir ut riktig verdi med «desimalar»-argumentet", {
+  )
+})
+
+test_that("num() ikke konverterer store tall til 
+          eksponentiell notasjon", {
```
Kva med små tal (som `1e-12`)?

statkarl commented on tests/testthat/test-rapporttekst-funksjoner.R:
```diff
@@ -113,3 +143,44 @@ test_that("akse_tall_format() gir ut en tom tekstvektor hvis den tar
+})
+
+test_that("prosent() gir ut riktig verdi når kun «x»-argumentet er gitt", {
+  expect_identical(prosent(numeric()), character(0))
```
Denne linja kunne godt vera skilt ut som éigen test, dvs. ein test på at funksjonen handterer inndata av lengd 0. Då får ein ei meir nyttig feilmelding.

Viss ein har ei naiv implementering av funksjonen som noko slikt (som fungerer nesten alltid):

```r
paste0(num(x), "\\prosent")
````

er det forvirrande å få ei feilmelding som seier at testen «prosent() gir ut riktig verdi når kun «x»-argumentet er gitt» feila. For funksjonen *vil* jo oftast fungera når kun `x`-argumentet er oppgjeve. Det er mykje betre å få ei feilmelding som seier at testen «prosent() gir ut riktig verdi ved tom innvektor» (elns.) feila.

Forresten inkonsistent at du skriv `character(0)` i staden for berre `character()`.

statkarl reviewed (changes requested):
Nokre kommentarar til.

statkarl commented on R/rapporttekst-funksjoner.R:
```diff
@@ -10,84 +10,90 @@
 #' @param x Vektor med tala ein ønskjer å få ut på LaTeX-format.
-#' @param desimalar Kor mange desimalar skal visast etter kommaet?
-#'   
+#' @param desimalar Kor mange desimalar som skal visast etter kommaet.
```
Kva tyder `NULL`-verdien for `desimalar`? (Det er jo til og med standardverdien!)

statkarl replied:
Ignorerer denne kommentaren. Ser det står forklart i detaljane.

orjanls committed:
Rettet grammatikkfeil og fjernet whitespace i dokumentasjon

orjanls committed:
Lagt til `nolint: object_name_linter` der det bør være camelCase

orjanls committed:
Forbedret testene for `prosent`

Forandret unødvendig hardkoding av forventet resultat
til å bruke `num` i stedet. Splitet en test i to for å
få nyttigere feilmelding dersom inndata er 0. Fjernet
tester som testet det samme, og forandret unødvendig
hardkoding av resultat fra `checkmate::check_numeric`.

orjanls committed:
Test for `num` at små tall ikke blir vist som eksponentiell notasjon

orjanls committed:
Fjernet unødvendige tester for `num` som testet det samme

orjanls committed:
Delte en test for `num` i to for mer spesifikk feilmelding

orjanls committed:
Endret `prosent` til å gi ut det samme som `num` dersom inndata er `NA`

orjanls committed:
Forbedret feilmelding for `expect_error` test av `prosent`

orjanls committed:
Forbedret feilmelding for `expect_error` test av `num`

orjanls committed:
Fjernet test som testet det samme og fjernet hardkoding av feilmelding

orjanls committed:
Endret feilmelding fra tester til å være på en linje

orjanls committed:
Forandret `prosent` til å ikke bruke `num` to ganger
